### PR TITLE
Bug fixes for timeout 

### PR DESCRIPTION
**Pull request by asn1809**
1. TImeout checks to include the check hook execution as well.
2. Incorporating review comments pending from #1649

### CODE REVIEW

File: internal/controller/util/json_util.go [hunk 1]
```diff
@@ -34,38 +35,71 @@ func EvaluateCheckHook(client client.Client, hook *kubeobjects.HookSpec, log log
...
 		resource := &corev1.Pod{}
 
-		err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
-		if err != nil {
-			return false, err
-		}
+		// err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
+		// if err != nil {
+		// 	return false, err
+		// }
```
Comment (nirs) [inline internal/controller/util/json_util.go]: Why do you keep this as commented code?

File: internal/controller/util/json_util.go [hunk 2]
```diff
@@ -34,38 +35,71 @@ func EvaluateCheckHook(client client.Client, hook *kubeobjects.HookSpec, log log
...
 
-		return EvaluateCheckHookExp(hook.Chk.Condition, resource)
+		//return EvaluateCheckHookExp(hook.Chk.Condition, resource)
```
Comment (nirs) [inline internal/controller/util/json_util.go]: Why do you keep this commented code?

File: internal/controller/util/json_util.go [hunk 3]
```diff
@@ -34,38 +35,71 @@ func EvaluateCheckHook(client client.Client, hook *kubeobjects.HookSpec, log log
...
 }
 
-func WaitUntilResourceExists(client client.Client, obj client.Object, nsScopedName types.NamespacedName,
+func EvaluateCheckHookExpWithTimeout(client client.Client, obj client.Object, nsScopedName types.NamespacedName,
+	timeout time.Duration, booleanExpression string,
+) (bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	ticker := time.NewTicker(pollInterval * time.Millisecond) // Poll every 100 milliseconds
```
Comment (nirs) [inline internal/controller/util/json_util.go]: This comment will quickly become stale when we change pollInternal. It is better to remove it.

However polling every 100 millisecond is too much. If we need to check a resource until the resource reach some state, the right way is to watch the resource. Each time the resources changes, you check the desired state.

Until we get a proper watch we can make this more efficient by using a backoff. We can start with short wait internal to be able to detect very short waits quickly (e.g. 10 ms), and double the interval on each retry, until we reach maximum internal (e.g. 10 seconds).

File: internal/controller/util/json_util.go [hunk 4]
```diff
@@ -232,27 +266,27 @@ func compareValues(val1, val2 interface{}, operator string) (bool, error) {
 	case float64:
 		v2, ok := val2.(float64)
 		if !ok {
-			return false, fmt.Errorf("mismatched types")
+			return false, fmt.Errorf("mismatched types %T and %T", val1, val2)
```
Comment (nirs) [inline internal/controller/util/json_util.go:224]: This error is not clear - which is the expected value? we can make it clear using:

    fmt.Errorf("type mismatch: expected: %T, actual: %T", expected, actual)

Same for other errors bellow.

**Comment by asn1809** [comment]
These changes are taken care of in another PR.

**Pull request closed without merging**